scripts/dts/gen_defines.py

```diff
@@ -177,6 +178,33 @@ def write_device_extern_header(device_header_out, edt):
         print("#endif /* DEVICE_EXTERN_GEN_H */", file=dev_header_file)
 
 
+def write_emul_extern_header(emul_header_out, edt):
+    # Generate header that will extern devicetree struct emuls
+
+    with open(emul_header_out, "w", encoding="utf-8") as emul_header_file:
+        print("#ifndef EMUL_EXTERN_GEN_H", file=emul_header_file)
+        print("#define EMUL_EXTERN_GEN_H", file=emul_header_file)
+        print("", file=emul_header_file)
+        print("#ifdef __cplusplus", file=emul_header_file)
+        print('extern "C" {', file=emul_header_file)
+        print("#endif", file=emul_header_file)
+        print("", file=emul_header_file)
+        print("/* Do not include directly. #include <zephyr/drivers/emul.h> instead. */",
+                file=emul_header_file)
+
+        for node in sorted(edt.nodes, key=lambda node: node.dep_ordinal):
+            print(f"extern const struct emul EMUL_DT_NAME_GET(DT_{node.z_path_id});" \
+                    f" /* dts_ord_{node.dep_ordinal} */",
+                    file=emul_header_file)
+
+        print("", file=emul_header_file)
+        print("#ifdef __cplusplus", file=emul_header_file)
+        print("}", file=emul_header_file)
+        print("#endif", file=emul_header_file)
+        print("", file=emul_header_file)
+        print("#endif /* EMUL_EXTERN_GEN_H */", file=emul_header_file)
```

Comment:
I don't think we should modify device infrastructure because of a subsystem design. If the emulator subsystem decided to be independent of devices (I still don't get the reason device model couldn't be re-used), it should be in all aspects, including any necessary tooling.

Reply:
If by "device infrastructure," you mean the build processes that operate on devicetrees, I don't think this modifies them significantly. This creates an internal-only header based on the devicetree contents, but we were already doing that for `struct device`. (A previous version of this PR just added to the already-existing header.) The `struct emul` declarations exposed in the header were not public previously, but right now, every `struct emul` definition is based on the devicetree. Basing the corresponding declarations on the devicetree seems like the natural way to go.

If we wanted to define `struct emul`s independently of the devicetree (as we can for `struct device` with `DEVICE_DEFINE`), and we wanted to make those declarations public, we would need to do it via some other mechanism. But to disassociate `struct emul`s from the devicetree entirely, we would need to remove/replace the existing mechanism for defining emulators, because it's based on the devicetree right now.

I think that emulators should probably continue to be based on the devicetree going forward. In any case, I think separating `struct emul` from devicetree is out of scope for this PR.

Reply:
@gmarull any more thoughts about this? The tool here is under the dts umbrella and `gen_defines.py` seems to fit since we're mirroring how devices work.

Emulators cannot use devices because they use the same DTS node as the device. A short example:

```
&i2c0 {
  my_dev: my_dev@80 {
    compatible = "test, my_dev";
    reg = <0x80>;
   ...
  };
};
```

This the above, we would create a `struct device` from that node in the driver logic. The emulator (when using the same DTS file) will need to create a parallel emulator that uses the same properties. If we were to use DEVICE_DT_DEFINE we would be creating a second `struct device` with the exact same name. Since both have to live in the same binary, we're using a `struct emul` instead but mirroring how devices work. Now, when the read device driver makes an i2c call, the emulated instance for `i2c0` above will know to route the call to the `struct emul` created for the same node.

If we were to split all the tooling for emulators out, we would be replicating a lot of the same code, since emulators are so tightly coupled to devices in DTS, I do not see a need for them to be done separately. Especially because any large changes to device tooling will also require changes to emulator tooling. It's just adding more liability and making it harder to maintain.

Reply:
Emulators are now independent of devices, it is a subsystem that can be disabled (not like devices for example). So from a design perspective, any of its requirements should also be independent, including scripts. I can't see any relevant function re-use in the current script really, so making it independent should be easy by keeping the same imports in a separate script. I also see that it now runs unconditionally, which is wrong. Why do I need to spend time generating `emul_extern.h` when I don't use emulators?

Keeping things independent eases maintenance as well. If I want to work on a device's script, I should not have to learn about emulators. A similar case happened in the past with PM. It was all mixed with devices, so anyone working on device.h had to learn about PM to make any serious change and not break things. Now we just keep a pointer and the rest is managed in the subsystem.

Reply:
I've posted https://github.com/zephyrproject-rtos/zephyr/pull/47798 as a way we can make this more generic and only have the EMUL externs come in when EMUL is being used.

Reply:
#47798 seems like it should work. Further comment there.

Reply:
@galak Should I be waiting for you to submit #47798, or did you want me to incorporate something like that into this PR?